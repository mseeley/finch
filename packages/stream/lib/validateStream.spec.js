/* eslint-env jest */

const { localNameOf } = require("@finch/core");
const validateStream = require("./validateStream");

describe(localNameOf(__filename), () => {
  const use = "module-id";
  const off = use;

  describe("when validating structure", () => {
    it("accepts composite stream", () => {
      const stream = [{ use }, [{ use }, [{ use }]]];

      expect(validateStream(stream)).toBeNull();
    });

    it("requires `use` or `off`", () => {
      expect(validateStream([{}])).toMatchSnapshot();

      expect(validateStream([{ use, off }])).toMatchSnapshot();
    });

    it("requires top-level array contains items", () => {
      expect(validateStream([])).toMatchSnapshot();
    });

    it("rejects additional properties on stream", () => {
      expect(validateStream([{ use, pink: "elephants" }])).toMatchSnapshot();
    });

    it("reports all errors", () => {
      expect(validateStream([{ use: "" }, { use: "" }])).toMatchSnapshot();
    });
  });

  describe("when validating stream `name`", () => {
    it("is optional", () => {
      expect(validateStream([{ use }])).toBeNull();
    });

    it("accepts a string", () => {
      expect(validateStream([{ use, name: "my-name" }])).toBeNull();
    });

    it("must have length", () => {
      expect(validateStream([{ use, name: "" }])).toMatchSnapshot();
    });
  });

  describe("when validating stream `off`", () => {
    it("is optional", () => {
      // `off` is optional when `use` is is provided.
      expect(validateStream([{ use }])).toBeNull();
    });

    it("accepts a string", () => {
      expect(validateStream([{ off }])).toBeNull();
    });

    it("must have length", () => {
      expect(validateStream([{ off: "" }])).toMatchSnapshot();
    });
  });

  describe("when validating stream `params`", () => {
    it("is optional", () => {
      expect(validateStream([{ use }])).toBeNull();
    });

    it("accepts an object", () => {
      expect(validateStream([{ use, params: { a: 1 } }])).toBeNull();
    });

    it("must have at least one property", () => {
      expect(validateStream([{ use, params: {} }])).toMatchSnapshot();
    });
  });

  describe("when validating stream `retryCount`", () => {
    it("is optional", () => {
      expect(validateStream([{ use }])).toBeNull();
    });

    it("accepts a number", () => {
      expect(validateStream([{ use, retryCount: 100 }])).toBeNull();
    });

    it("must be >= 1", () => {
      expect(validateStream([{ use, retryCount: 1 }])).toBeNull();
      expect(validateStream([{ use, retryCount: 0 }])).toMatchSnapshot();
    });
  });

  describe("when validating stream `retryWait`", () => {
    it("is optional", () => {
      expect(validateStream([{ use }])).toBeNull();
    });

    it("accepts a number", () => {
      expect(validateStream([{ use, retryWait: 100 }])).toBeNull();
    });

    it("must be >= 0", () => {
      expect(validateStream([{ use, retryWait: 0 }])).toBeNull();
      expect(validateStream([{ use, retryWait: -1 }])).toMatchSnapshot();
    });
  });

  describe("when validating stream `use`", () => {
    it("is optional", () => {
      // `use` is optional when `off` is is provided.
      expect(validateStream([{ off }])).toBeNull();
    });

    it("accepts a string", () => {
      expect(validateStream([{ use }])).toBeNull();
    });

    it("must have length", () => {
      expect(validateStream([{ use: "" }])).toMatchSnapshot();
    });
  });

  describe("when validating stream `verbose`", () => {
    it("is optional", () => {
      expect(validateStream([{ use }])).toBeNull();
    });

    it("accepts a boolean", () => {
      expect(validateStream([{ use, verbose: true }])).toBeNull();
    });

    it("must be boolean", () => {
      expect(validateStream([{ use, verbose: "true" }])).toMatchSnapshot();
    });
  });
});
