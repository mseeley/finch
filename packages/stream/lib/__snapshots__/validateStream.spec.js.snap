// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`stream/lib/validateStream.spec when validating stream \`name\` must have length 1`] = `
Array [
  Object {
    "errors": Array [
      "should NOT be shorter than 1 characters",
    ],
    "pointer": "/0/name",
  },
]
`;

exports[`stream/lib/validateStream.spec when validating stream \`off\` must have length 1`] = `
Array [
  Object {
    "errors": Array [
      "should NOT be shorter than 1 characters",
    ],
    "pointer": "/0/off",
  },
]
`;

exports[`stream/lib/validateStream.spec when validating stream \`params\` must have at least one property 1`] = `
Array [
  Object {
    "errors": Array [
      "should NOT have fewer than 1 properties",
    ],
    "pointer": "/0/params",
  },
]
`;

exports[`stream/lib/validateStream.spec when validating stream \`retryCount\` must be >= 1 1`] = `
Array [
  Object {
    "errors": Array [
      "should be >= 1",
    ],
    "pointer": "/0/retryCount",
  },
]
`;

exports[`stream/lib/validateStream.spec when validating stream \`retryWait\` must be >= 0 1`] = `
Array [
  Object {
    "errors": Array [
      "should be >= 0",
    ],
    "pointer": "/0/retryWait",
  },
]
`;

exports[`stream/lib/validateStream.spec when validating stream \`use\` must have length 1`] = `
Array [
  Object {
    "errors": Array [
      "should NOT be shorter than 1 characters",
    ],
    "pointer": "/0/use",
  },
]
`;

exports[`stream/lib/validateStream.spec when validating stream \`verbose\` must be boolean 1`] = `
Array [
  Object {
    "errors": Array [
      "should be boolean",
    ],
    "pointer": "/0/verbose",
  },
]
`;

exports[`stream/lib/validateStream.spec when validating structure rejects additional properties on stream 1`] = `
Array [
  Object {
    "errors": Array [
      "should NOT have additional properties: pink",
    ],
    "pointer": "/0",
  },
]
`;

exports[`stream/lib/validateStream.spec when validating structure reports all errors 1`] = `
Array [
  Object {
    "errors": Array [
      "should NOT be shorter than 1 characters",
    ],
    "pointer": "/0/use",
  },
  Object {
    "errors": Array [
      "should NOT be shorter than 1 characters",
    ],
    "pointer": "/1/use",
  },
]
`;

exports[`stream/lib/validateStream.spec when validating structure requires \`use\` or \`off\` 1`] = `
Array [
  Object {
    "errors": Array [
      "should have required property 'off'",
      "should have required property 'use'",
      "should match exactly one schema in oneOf",
    ],
    "pointer": "/0",
  },
]
`;

exports[`stream/lib/validateStream.spec when validating structure requires \`use\` or \`off\` 2`] = `
Array [
  Object {
    "errors": Array [
      "should match exactly one schema in oneOf",
    ],
    "pointer": "/0",
  },
]
`;

exports[`stream/lib/validateStream.spec when validating structure requires top-level array contains items 1`] = `
Array [
  Object {
    "errors": Array [
      "should NOT have fewer than 1 items",
    ],
    "pointer": "",
  },
]
`;
